packages/data/addon/adapters/bard-facts-v2.ts

```diff
@@ -0,0 +1,406 @@
+ * @param {Parameters} parameters - parameters object
+ * @param {boolean} includeFieldProj - whether to include the projected field in the formatted name e.g. age|id  OR  age
+ * @returns {string} formatted dimension name
+ */
+function formatDimensionFieldName(field: string, parameters: Parameters, includeFieldProj = true): string {
+  const [fieldName, displayField = 'id'] = field.split('.'); // default dimension field projection to "id"
+  let parameterString = serializeParameters(parameters);
+
+  if (parameterString.length > 0) {
+    parameterString = `(${parameterString})`;
+  }
+  return `${fieldName}${parameterString}${includeFieldProj ? '|' + displayField : ''}`;
+}
+
+/**
+ * Serializes a list of filters to a fili filter string
+ * @param {Array<Filter>} filters - list of filters to be ANDed together for fili
+ * @param {Filter} filter - filters object
+ */
+export function serializeFilters(filters: Filter[]) {
+  return filters
+    .map(filter => {
+      const { field, operator, values, parameters } = filter;
+      const serializedValues = values
+        .map(v => String(v).replace(/"/g, '""')) // csv serialize " -> ""
+        .map(v => `"${v}"`) // wrap each "value"
+        .join(','); // comma to separate
+      const formattedFieldName = formatDimensionFieldName(field, parameters, true);
+
+      return `${formattedFieldName}-${operator}[${serializedValues}]`;
+    })
+    .join(',');
+}
+
+export default class BardFactsAdapter extends EmberObject {
+  /**
+   * @property namespace
+   */
+  namespace = 'v1/data';
+
+  /**
+   * @property {Service} ajax
+   */
+  @service ajax: TODO;
+
+  /**
+   * Builds the dimensions path for a request
+   *
+   * @method _buildDimensionsPath
+   * @private
+   * @param {Object} request
+   * @return {String} dimensions path
+   */
+  _buildDimensionsPath(request: Request /*options*/): string {
+    const dimensions = array(request.columns)
+      .filterBy('type', 'dimension')
+      .map(dim => formatDimensionFieldName(dim.field, dim.parameters, false));
+
+    return dimensions.length
+      ? `/${array(dimensions)
+          .uniq()
+          .join('/')}`
+      : '';
+  }
+
+  /**
+   * Builds a dateTime param string for a request
+   *
+   * @method _buildDateTimeParam
+   * @private
+   * @param {Object} request
+   * @return {String} dateTime param value
+   */
+  _buildDateTimeParam(request: Request): string {
+    const dateTimeFilters = request.filters.filter(fil => fil.field === 'dateTime');
+
+    return dateTimeFilters.map(filter => `${filter.values[0]}/${filter.values[1]}`).join(',');
+  }
+
+  /**
+   * Builds a metrics param string for a request
+   *
+   * @method _buildMetricsParam
+   * @private
+   * @param {Object} request
+   * @return {String} metrics param value
+   */
+  _buildMetricsParam(request: Request): string {
+    const metrics = request.columns.filter(col => col.type === 'metric');
+    const metricIds = (metrics || []).map(metric =>
```

Comment:
Shouldn't metrics always be defined if columns is defined?
```suggestion
    const metricIds = metrics.map(metric =>
```